ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -112,29 +93,68 @@ let create ~__context ~cluster ~host ~pif =
-      match result with
-      | Result.Ok () ->
-        Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF:pifref ~enabled:true
-          ~current_operations:[] ~allowed_operations:[] ~other_config:[];
-        debug "Cluster_host.create was successful; cluster_host: %s" (Ref.string_of ref);
-        ref
-      | Result.Error error ->
-        warn "Error occurred during Cluster_host.create";
-        handle_error error
-    )
+      let pifrec = Db.PIF.get_record ~__context ~self:pif in
+      assert_pif_prerequisites (pif,pifrec);
+      Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF:pif ~enabled:false
+        ~current_operations:[] ~allowed_operations:[] ~other_config:[] ~joined:false;
+      ref
+  )
+
+(* Enable cluster_host in client layer via clusterd *)
+let resync_host ~__context ~host =
+  match (find_cluster_host ~__context ~host) with
+  | None      -> () (* no clusters exist *)
+  | Some self ->    (* cluster_host and cluster exist *)
+    with_clustering_lock (fun () ->
+        (* Join cluster host in clusterd *)
+        if not (Db.Cluster_host.get_joined ~__context ~self) then begin
+          debug "Enabling clusterd and joining cluster_host %s" (Ref.string_of self);
+          Xapi_clustering.Daemon.enable ~__context;
+          let dbg = Context.string_of_task __context in
+          let cluster = Db.Cluster_host.get_cluster ~__context ~self in
+          let cluster_token = Db.Cluster.get_cluster_token ~__context ~self:cluster in
+          let pifref = Db.Cluster_host.get_PIF ~__context ~self in
+          let pifrec = Db.PIF.get_record ~__context ~self:pifref in
+          let ip = ip_of_pif (pifref,pifrec) in
+          let ip_list = List.map (fun self ->
+            let pref = Db.Cluster_host.get_PIF ~__context ~self in
+            let prec = Db.PIF.get_record ~__context ~self:pref in
+              ip_of_pif (pref,prec)
+            ) (Db.Cluster.get_cluster_hosts ~__context ~self:cluster) in
+          let result = Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list in
+          match result with
+          | Result.Ok () ->
+            Db.Cluster_host.set_joined ~__context ~self ~value:true;
+            Db.Cluster_host.set_enabled ~__context ~self ~value:true;
+            debug "Cluster_host.create was successful; cluster_host: %s" (Ref.string_of self);
+          | Result.Error error ->
+            warn "Error occurred during Cluster_host.create";
+            handle_error error
+        end;
+        if Db.Cluster_host.get_enabled ~__context ~self then begin
+        (* [enable] unconditionally invokes low-level enable operations and is idempotent. *)
+          (* RPU reformats partition, losing service status, never re-enables clusterd *)
+          debug "Cluster_host %s is enabled, starting up xapi-clusterd" (Ref.string_of self);
+          Xapi_clustering.Daemon.enable ~__context;
+
+          (* Best-effort Cluster_host.enable *)
```

Comment:
This doesn't seem to be best-effort anymore, if it fails the whole function fails. Thats fine, just remove the comment.